src/components/UriViewer/LogViewer.ts

```diff
@@ -0,0 +1,224 @@
+ * @member logs - An array of LogInfo objects.
+ */
+export type LogViewerProps = {
+  modalTitle: string;
+  logs: LogInfo[];
+  onDismiss: () => void;
+};
+
+/**
+ * Props for the SimpleTab component. See src/components/tabBars.js
+ */
+type SimpleTabProps = {
+  key: string;
+  title: string;
+  width: number;
+};
+
+type FetchedLogData = {
+  textContent: string | undefined; // The literal characters of the log file.
+  downloadUri: string | undefined; // The URI to use for downloading the log file. May or may not have token appended, depending on if the file is public or private.
+};
+
+const modalMaxWidth = 1100;
+const tabMaxWidth = modalMaxWidth / 4 - 20;
+
+export const LogViewer = _.flow(withDisplayName('LogViewer'))(({ modalTitle, logs, onDismiss }: LogViewerProps) => {
+  const [currentlyActiveLog, setCurrentlyActiveLog] = _.isEmpty(logs)
+    ? useState<LogInfo | undefined>(undefined)
+    : useState<LogInfo | undefined>(logs[0]);
+
+  // string = fetched log content, undefined = loading, null = error.
+  const [activeTextContent, setActiveTextContent] = useState<string | undefined | null>(undefined);
+  const [activeDownloadUri, setActiveDownloadUri] = useState<string | undefined>(undefined);
+  const signal = useCancellation();
+  const fetchLogContent = useCallback(
+    async (azureBlobUri: string): Promise<FetchedLogData | null> => {
+      if (!isAzureUri(azureBlobUri)) {
+        return null;
+      }
+      try {
+        const response = await Ajax(signal).AzureStorage.blobMetadata(azureBlobUri).getData();
+        const uri = _.isEmpty(response.azureSasStorageUrl) ? response.azureStorageUrl : response.azureSasStorageUrl;
+        return { textContent: response.textContent, downloadUri: uri };
+      } catch (e) {
+        return null;
+      }
+    },
+    [signal]
+  );
+
+  useOnMount(() => {
+    fetchLogContent(logs[0].logUri).then((content) => {
+      if (_.isEmpty(content?.textContent)) {
+        setActiveTextContent(null);
+      } else {
+        setActiveTextContent(content?.textContent);
+      }
+      setActiveDownloadUri(content?.downloadUri);
+    });
+  });
+
+  useEffect(() => {
+    const fetch = async (logUri: string) => {
+      const res = await fetchLogContent(logUri);
+      if (_.isEmpty(res?.textContent)) {
+        setActiveTextContent(null); // Fetch failed. Display error.
+      } else {
+        setActiveTextContent(res?.textContent);
+      }
+      setActiveDownloadUri(res?.downloadUri);
+    };
+
+    // when switching tabs, reset content to undefined while we fetch the new content.
+    setActiveTextContent(undefined);
+    setActiveDownloadUri(undefined);
+
+    // tab switching set the currently active log (which triggers this effect). Fetch the content for the new log.
+    const uri = currentlyActiveLog?.logUri;
+    if (!_.isEmpty(uri)) {
+      fetch(uri);
+    }
+  }, [logs, currentlyActiveLog, fetchLogContent]);
+
+  const renderActiveTextContent = () => {
```

Comment:
It might be worth looking at the LoadedState type in packages/core-utils for making the current state more clear. As is, `undefined` means "loading" and `null` means "error" is a distinction that's easy to miss.